internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,256 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type Watcher interface {
+	Watch(source.Source) error
+}
+
+// Cache is a storage mechanism for keeping track of
+// managed content sources. It also exposes methods
+// for retrieving the stored references of the managed
+// content
+type Cache interface {
+	io.Closer
+	// Watch establishes watches for all provided client.Objects
+	Watch(context.Context, Watcher, ...client.Object) error
+}
+
+// CloserSyncingSource is a wrapper of the controller-runtime
+// source.SyncingSource that includes methods for:
+//   - Getting a specific resource synced by the source
+//   - Listing resources synced by the source
+//   - Closing the source, stopping it's interaction with the Kubernetes API server and reaction to events
+type CloserSyncingSource interface {
+	source.SyncingSource
+	io.Closer
+}
+
+type Sourcerer interface {
+	Source(schema.GroupVersionKind, client.Object, func(context.Context)) (CloserSyncingSource, error)
+}
+
+type cache struct {
+	sources     map[schema.GroupVersionKind]CloserSyncingSource
+	sourcerer   Sourcerer
+	owner       client.Object
+	syncTimeout time.Duration
+	mu          sync.Mutex
+}
+
+func NewCache(sourcerer Sourcerer, owner client.Object, syncTimeout time.Duration) Cache {
+	return &cache{
+		sources:     make(map[schema.GroupVersionKind]CloserSyncingSource),
+		sourcerer:   sourcerer,
+		owner:       owner,
+		syncTimeout: syncTimeout,
+	}
+}
+
+var _ Cache = (*cache)(nil)
+
+func (c *cache) Watch(ctx context.Context, watcher Watcher, objs ...client.Object) error {
+	if watcher == nil {
+		return errors.New("nil watcher provided")
+	}
+
+	gvkSet, err := gvksForObjects(objs...)
+	if err != nil {
+		return fmt.Errorf("getting set of GVKs for managed objects: %w", err)
+	}
+
+	if err := c.removeStaleSources(gvkSet); err != nil {
+		return fmt.Errorf("removing stale sources: %w", err)
+	}
+	return c.startNewSources(ctx, gvkSet, watcher)
+}
+
+func (c *cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errs := []error{}
+	for _, source := range c.sources {
+		if err := source.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func (c *cache) startNewSources(ctx context.Context, gvks sets.Set[schema.GroupVersionKind], watcher Watcher) error {
+	cacheGvks := sets.New[schema.GroupVersionKind]()
+	for gvk := range c.getSources() {
+		cacheGvks.Insert(gvk)
+	}
+	gvksToCreate := gvks.Difference(cacheGvks)
+
+	errs := make(chan error)
+	wg := sync.WaitGroup{}
+	for _, gvk := range gvksToCreate.UnsortedList() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := c.startNewSource(ctx, gvk, watcher)
+			if err != nil {
+				errs <- err
+			}
+		}()
+	}
+
+	// watch for errors coming through the error channel and append them to the
+	// list of source errors until the error channel is closed
+	errWriteWg := sync.WaitGroup{}
+	errWriteWg.Add(1)
+	sourcesErrors := []error{}
+	go func() {
```

Comment:
I think we can do this part synchronously if we run `wg.Wait()` and `close(errs)` in a separate go routine, right after we finish the startSources goroutine loop.

Reply:
addressed in https://github.com/operator-framework/operator-controller/pull/1119/commits/af075c5137065a5e0b2e0cfc1b4debbb367f792f